server/mediaserver.go

```diff
@@ -407,25 +437,72 @@ func gotRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 								errFunc("Download", url, err)
 								return
 							}
+							hash := crypto.Keccak256(data)
+							SegHashLock.Lock()
+							segHashes[i] = hash
+							SegHashLock.Unlock()
+
 							sid, err := parseStreamID(url)
 							if err != nil {
 								errFunc("StreamID", url, err)
 								return
 							}
+							glog.V(common.DEBUG).Infof("legacy downloaded sid %s seg name %s", sid, parseSegName(url))
 							newSeg := &stream.HLSSegment{SeqNo: seg.SeqNo, Name: parseSegName(url), Data: data, Duration: seg.Duration}
-							s.LivepeerNode.VideoSource.InsertHLSSegment(sid, newSeg)
+							cpl := s.CurrentPlaylist
+							if cpl != nil {
+								cpl.InsertHLSSegment(jid, sid, newSeg)
+							}
+						}
 
+						dlFromOSFunc := func(turi *net.TypedURI, i int) {
+							defer func() {
+								cond.L.Lock()
+								n--
+								if n == 0 {
+									cond.Signal()
+								}
+								cond.L.Unlock()
+							}()
+							glog.V(common.DEBUG).Infof("i: %d, start downloading segment from %s sid %s seg name %s", i,
+								turi.Storage, turi.StreamID, parseSegNameSimple(turi.Uri))
+							data, err := drivers.GetSegmentData(turi)
+							if err != nil {
+								errFunc("Download", turi.Uri, err)
+								return
+							}
+							glog.V(common.DEBUG).Infof("segment downloaded %d bytes from %s sid %s seg name %s", len(data),
+								turi.Storage, turi.StreamID, parseSegNameSimple(turi.Uri))
 							hash := crypto.Keccak256(data)
 							SegHashLock.Lock()
 							segHashes[i] = hash
 							SegHashLock.Unlock()
-
+							newSeg := &stream.HLSSegment{SeqNo: seg.SeqNo,
+								Name: parseSegNameSimple(turi.Uri),
+								Data: data, Duration: seg.Duration}
+							cpl := s.CurrentPlaylist
+							if cpl != nil {
+								cpl.InsertHLSSegment(jid, core.StreamID(turi.StreamID), newSeg)
+							}
 						}
-
+						glog.V(common.DEBUG).Infof("Got result from transcode: %+v", res)
 						for i, v := range res.Segments {
-							go dlFunc(v.Url, i)
+							if v.Turl == nil {
+								// old orchestrator
+								go dlFunc(v.Url, i)
+							} else {
+								cpl := s.CurrentPlaylist
+								if cpl != nil {
+									inserted, err := cpl.InsertLink(jid, seg.SeqNo, v.Turl, seg.Duration)
```

Comment:
if InsertLink succeeds, then this deadlocks since the segment hashes are never set. Will fix.